ingestion-beam/src/main/java/com/mozilla/telemetry/avro/GuidedJsonDecoder.java

```diff
@@ -0,0 +1,410 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package com.mozilla.telemetry.avro;
+
+import com.fasterxml.jackson.core.JsonFactory;
+import com.fasterxml.jackson.core.JsonParser;
+import com.fasterxml.jackson.core.JsonToken;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.util.TokenBuffer;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Stack;
+
+import org.apache.avro.AvroTypeException;
+import org.apache.avro.Schema;
+import org.apache.avro.io.ParsingDecoder;
+import org.apache.avro.io.parsing.JsonGrammarGenerator;
+import org.apache.avro.io.parsing.Parser;
+import org.apache.avro.io.parsing.Symbol;
+import org.apache.avro.util.Utf8;
+
+/**
+ * A custom JSON decoder based on the Parsing JsonDecoder in the
+ * `org.apache.avro.io.JsonDecoder` package.
+ *
+ * <p>The default JsonDecoder (as of Avro 1.8.2) does support ingestion of JSON
+ * documents where casting of values are inferred from the provided schema (see
+ * AVRO-1582 [1]). Instead, the decoder expects values that are emitted by the
+ * JsonEncoder in the form `{"NAME": {"TYPE": "VALUE"}}`. While this enables a
+ * lossless type-encoding when using JSON as a transport, it is inconvenient for
+ * ingesting data that are directly encoded into JSON i.e. values in the form
+ * `{"NAME": "VALUE"}`.
+ *
+ * <p>This decoder is designed as a companion to the jsonschema-transpiler [2],
+ * which emits valid Avro schemas from JSON Schema. These schemas are used to
+ * guide the serialization process. This is required to disambiguate floating
+ * points from integers and records from maps. To support BigQuery, the decoder
+ * uses unions to enforce required fields and will fail on variant-types.
+ * Additionally, field names are renamed to match the column-name rules for
+ * BigQuery tables. The source JSON schemas can be underspecified or overly
+ * complex. In such cases, the transpiler will cast fields into strings. This
+ * encoder supports encoding tree-nodes into strings when guided by the provided
+ * schema.
+ *
+ * <p>This class shares some code from the JsonDecoder class. The original class
+ * cannot be extended outside of the Avro package. The algorithm for buffering
```

Comment:
It's unsavory, but we've done it before: we can define this class inside the avro package so that it's able to inherit from avro's JsonDecoder. If that removes a significant amount of copied code here, I'd support it. As long as we document in the javadoc why we're doing what we're doing.

Reply:
Are there any pointers to defining a class within an external package? On another hand,I re-read through the variable scoping in the JsonDecoder class and found out that the JsonParser `in` that drives the decoding is a private variable. This means there's no way around this other than forking or re-implementing from the ParsingDecoder class.

Extending the class would reduce the code to`doAction`, `readString`, and `readIndex`. The `doAction` method needs to be re-implemented because of the custom renaming rule for records. One implementation outside of apache that extends the JsonDecoder is [PIMCO/avro](https://github.com/PIMCO/avro). They support a few custom extensions like decimal types. It looks like their fork changes the private variable into a protected one, which comes at the cost of maintaining a fork of the project.

Reply:
> Are there any pointers to defining a class within an external package?

It's rather silly that it works this way, but if you change the package declaration at the top to `org.apache.avro` (or whichever subpackage you need) and move the file to:

    ingestion-beam/src/main/java/org/apache/avro/GuidedJsonDecoder.java 

Then it gets all the privileges as if it had been defined in the avro source code.

Reply:
Assuming that tricks to inherit won't actually work due to the private `in`, can you document in the source which methods are modified? Perhaps every modified gets javadoc that specifies it's our code?

Reply:
I'll annotate the methods that contain mostly custom code. I did rewrite using the original as reference, so there are quite a few cosmetic changes to variable names and specific error handling.